Make presented beatmap a recommended one

LittleEndu opened the pull request:
Closes #8732
Closes #11062 (I assume it's fixed by DifficultyRecommender getting online faster)

There's a lot of small things this PR does:
 * Move ``DifficultyRecommender`` from ``SongSelect`` to ``OsuGameBase``
   * ``SongSelect`` makes use of it via injection
   * This *does not* move it from its current namespace
 * Rewrite ``DifficultyRecommender`` to recommend from all rulesets
   * This used to only recommend from current ruleset
   * Now orders all rulesets, based on which recommend higher star ratings, and tries to recommend something from them in order, while current ruleset is always recommended first
   * This shouldn't change ``CarouselBeatmapSet`` behaviour which already filtered out beatmaps before asking for recommendation 
   * Ordering of rulesets is set to happen only once every ruleset change. Probably could be optimized better :/ (probably want to order them once and just change the order of current ruleset in function scope only. Could make use of OrderedDictionary then)
 * Make ``PresentBeatmap()`` use new ``DifficultyRecommender``
   * First ``difficultyPredicate`` isn't used to find a single beatmap anymore. Instead it's used to filter all beatmaps from the ``BeatmapSetInfo``
   * Those filtered beatmaps are then fed to ``DifficultyRecommender`` and the recommended beatmap is presented
   * If ``DifficultyRecommender`` fails, first try to see if the predicate matched anything and present that, second try to use default predicate (current ruleset) and present that, if all fails just present first beatmap

Even before this PR, ``PresentBeatmap`` by-passed any user selection (``CarouselGroupEagerSelect`` tries to keep track of ``lastSelectedIndex``) and would have always presented the easiest difficulty. This is now swapped with presenting the recommended difficulty.

LittleEndu committed:
Move DifficultyRecommender to OsuGameBase

LittleEndu committed:
Recommend from all rulesets

LittleEndu committed:
Reset best order on ruleset change

LittleEndu committed:
Present recommended beatmaps

LittleEndu committed:
Offline fallback and commenting

LittleEndu committed:
Better fallback logic

bdach commented on osu.Game/Screens/Select/DifficultyRecommender.cs:
```diff
@@ -72,6 +85,26 @@ private void calculateRecommendedDifficulties()
+                return bestRulesetOrder;
+
+            var otherRulesets = recommendedStarDifficulty.ToList()
+                                                         .Where(pair => !pair.Key.Equals(ruleset.Value))
```
To achieve the effect I assume is desired here, you could actually go a little weird to make this shorter and leverage `OrderBy`:
```csharp
> new[] { false, false, true, false }.OrderByDescending(x => x)
OrderedEnumerable<bool, bool> { true, false, false, false }
```
and then chain with `ThenByDescending` to get the secondary ordering.

bdach commented on osu.Game/OsuGame.cs:
```diff
@@ -346,11 +344,24 @@ public void PresentBeatmap(BeatmapSetInfo beatmap, Predicate<BeatmapInfo> diffic
+                    if (difficultyCriteria != null)
+                        selection = beatmaps.First();
+                    else
+                        selection = databasedSet.Beatmaps.Find(b => b.Ruleset.Equals(Ruleset.Value)) ?? databasedSet.Beatmaps.First();
```
Not sure this discarding `difficultyCriteria` to only check ruleset is correct. Sounds like it might break some previous changes in this area.

LittleEndu replied:
the first path of the if statement would use the beatmaps filtered by ``difficultyCriteria``

bdach reviewed (changes requested):
At this level of complexity of the logic in here tests are a must for me.

LittleEndu commented:
> tests are a must for me.
Everything in here is already tested by previously written tests. New functionality is tied to online as recommendations are based on performance points.

You might be able to overwrite the cached ``DifficultyRecommender`` in tests and make it not rely on online and then test it. Not sure if this is how DI works.

bdach commented:
We actually have mocking for online now since #8714 which might come in handy. Ideally I'd like to see the ruleset sorting thing tested.

LittleEndu commented:
I think I can write a test for it then. Seems simple enough, just need ``DummyAPIAccess``'s ``HandleRequest`` to return a dummy user with dummy statistics when ``DifficultyRecommender`` fires a ``GetUserRequest`` request. 

I'll look into it tomorrow.

LittleEndu commented on osu.Game/Screens/Select/DifficultyRecommender.cs:
```diff
@@ -44,16 +46,27 @@ private void load()
             {
-                return beatmaps.OrderBy(b =>
+                if (!recommendedStarDifficulty.TryGetValue(ruleset.Value, out var stars))
+                    break;
```
This should probably be ``continue;`` instead or shouldn't be here at all.

LittleEndu commented on osu.Game/OsuGame.cs:
```diff
@@ -346,11 +344,24 @@ public void PresentBeatmap(BeatmapSetInfo beatmap, Predicate<BeatmapInfo> diffic
+                // Find beatmaps that match our predicate.
+                var beatmaps = databasedSet.Beatmaps.Where(b => difficultyCriteria?.Invoke(b) ?? true);
+                if (!beatmaps.Any())
+                    beatmaps = databasedSet.Beatmaps;
```
This should write null to ``difficultyCriteria`` as it wasn't able to provide any difficulties and shouldn't be used again. Would fix the comment below.

LittleEndu replied:
made it use difficulty criteria only once so this shouldn't be needed.

LittleEndu committed:
Simplify recommended beatmap presenting

LittleEndu committed:
Calculate best ruleset order only once

LittleEndu committed:
Very basic testing

LittleEndu commented:
Needs tests for when current ruleset doesn't have difficulties that the next ruleset recommended is the one player has higher recommended star difficulty in

LittleEndu commented on osu.Game/Screens/Select/DifficultyRecommender.cs:
```diff
@@ -90,19 +88,22 @@ private void calculateRecommendedDifficulties()
 
-            var rulesetList = new List<RulesetInfo>(new[] { ruleset.Value });
-            rulesetList.AddRange(otherRulesets);
+            return moveCurrentRulesetToFirst();
```
Both return statements return the same thing, can make if statement into ``bestRulesetOrder == null``

LittleEndu commented on osu.Game/Screens/Select/DifficultyRecommender.cs:
```diff
@@ -32,8 +32,6 @@ public class DifficultyRecommender : Component, IOnlineComponent
         {
             api.Register(this);
-
-            ruleset.ValueChanged += _ => bestRulesetOrder = null;
```
``bestRulesetOrder`` should instead be reset when new values are written to ``recommendedStarDifficulty`` when the request succeeds.

LittleEndu committed:
Optimize recommender (for custom rulesets)

LittleEndu committed:
More testing

LittleEndu commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs:
```diff
@@ -0,0 +1,168 @@
+            // Make sure we are on standard ruleset
+            presentAndConfirm(osuImport, 0, 1);
+
+            // Present mixed difficulty set, expect ruleset with highest star difficulty
```
Should say second highest as the highest is missing in this test

LittleEndu commented on osu.Game/OsuGameBase.cs:
```diff
@@ -99,6 +100,9 @@ public class OsuGameBase : Framework.Game, ICanAcceptFiles
 
         public bool IsDeployedBuild => AssemblyVersion.Major > 0;
 
+        [Cached]
+        protected readonly DifficultyRecommender DifficultyRecommender = new DifficultyRecommender();
```
I put the recommender in base game so that only one instance is active. That's to lower API calls, I don't know how expensive calling ``GetUserRequest`` 4 times is but I can't imagine it being cheap.

LittleEndu replied:
I'm thinking about again working on this PR over the weekend. I get all the other comments but this one I still would like clarified.

> resolve at necessary call sites 

I think in this PR the test, ``SongSelect``, and ``OsuGame`` are the only call sites. But the fact that ``OsuGame.PresentBeatmap`` needs access to recommender to present recommended beatmaps means that ``DifficultyRecommender`` needs to be accessed at base level. To me it made sense to cache it at base level too and you get reduction to API calls as a bonus.

bdach replied:
The long and short of it is that `OsuGame` and `OsuGameBase` should be as slim as they could possibly hope to be. If this can't be placed anywhere else without significant overhead, so be it, but those classes *need* to remain manageable.

LittleEndu replied:
This also helps ``DifficultyRecommender`` get online faster to start making recommendations from the start, fix for #11062

LittleEndu committed:
Fix CI complaints

bdach commented:
Is this at a reviewable state right now? Hard to tell with the changes coming in.

LittleEndu commented:
It should be. Can't think of what else would need testing. Would like to figure out why testing in visual browser doesn't work, but I have no idea what broke it.

smoogipoo commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -104,14 +102,13 @@ public abstract class SongSelect : OsuScreen, IKeyBindingHandler<GlobalAction>
 
         [BackgroundDependencyLoader(true)]
-        private void load(AudioManager audio, DialogOverlay dialog, OsuColour colours, SkinManager skins, ScoreManager scores)
+        private void load(AudioManager audio, DialogOverlay dialog, OsuColour colours, SkinManager skins, ScoreManager scores, DifficultyRecommender recommender)
```
Should be fine so long as a reference isn't being held out-in (i.e. object with longer lifetime holding reference to object with a shorter lifetime).

As far as I can see that's not the case here.

bdach commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs:
```diff
@@ -0,0 +1,168 @@
+                    {
+                        case GetUserRequest userRequest:
+
+                            decimal pp = userRequest.Ruleset.ID switch
+                            {
+                                0 => 336, // Expected recommended star difficulty 2*
+                                1 => 928, // Expected recommended star difficulty 3*
+                                2 => 1905, // Expected recommended star difficulty 4*
+                                3 => 3329, // Expected recommended star difficulty 5*
+                                _ => 0
+                            };
```
We don't generally use switch expressions anywhere, prefer standard ones.

bdach commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs:
```diff
@@ -0,0 +1,168 @@
+            {
+                ((DummyAPIAccess)API).HandleRequest = req =>
+                {
+                    switch (req)
+                    {
+                        case GetUserRequest userRequest:
+
+                            decimal pp = userRequest.Ruleset.ID switch
+                            {
+                                0 => 336, // Expected recommended star difficulty 2*
+                                1 => 928, // Expected recommended star difficulty 3*
+                                2 => 1905, // Expected recommended star difficulty 4*
+                                3 => 3329, // Expected recommended star difficulty 5*
+                                _ => 0
+                            };
+
+                            userRequest.TriggerSuccess(new User
+                            {
+                                Username = @"Dummy",
+                                Id = 1001,
+                                Statistics = new UserStatistics
+                                {
+                                    PP = pp
+                                }
+                            });
+                            break;
+                    }
```
Way way *way* nested. Could factor out a method or two here.

bdach commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs:
```diff
@@ -0,0 +1,168 @@
+
+            for (int i = 0; i < 5; i++)
+            {
+                importFunctions.Add(importBeatmap(i, new List<RulesetInfo> { null, null, null, null, null }));
```
`Enumerable.Repeat` exists, and even then, this is quite strange what with passing `null` five times for it to only fallback to the first available ruleset at `importBeatmap`.

bdach commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs:
```diff
@@ -0,0 +1,168 @@
+        [Test]
+        public void TestBestRulesetIsRecommended()
+        {
+            var osuRuleset = rulesets.AvailableRulesets.First(r => r.ID == 0);
```
Generally these should be `Single`s, but you can actually simply do
```suggestion
            var osuRuleset = rulesets.GetRuleset(0);
```
(not going to point out every single instance).

bdach commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs:
```diff
@@ -0,0 +1,168 @@
+            return () => imported;
+        }
+
+        private void presentAndConfirm(Func<BeatmapSetInfo> getImport, int importedID, int expextedDiff)
```
```suggestion
        private void presentAndConfirm(Func<BeatmapSetInfo> getImport, int importedID, int expectedDiff)
```
(+ usages)

bdach commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs:
```diff
@@ -0,0 +1,168 @@
+            presentAndConfirm(mixedImport, 1, 2);
+        }
+
+        private Func<BeatmapSetInfo> importBeatmap(int importID, List<RulesetInfo> rulesets)
```
Not sure I like the way this is written too much. A little complicated to parse but I'm not sure if it's just me.

Would probably pre-construct the beatmap info in the first step of each test, store them to a variable/field and then operate on the created info through `Func`s in the asserts (so pass the info instead of passing magic indices to `presentAndConfirm`).

LittleEndu replied:
Seems like you resolved this yourself in 2913a8183538f823858aace50834f9e38a74b726 but I think that code was copied from [TestScenePresentBeatmap](https://github.com/ppy/osu/blob/93269606cd78c1193d561ba649f254afa1255c27/osu.Game.Tests/Visual/Navigation/TestScenePresentBeatmap.cs#L88-L129). Maybe something over there needs to be changed too

bdach replied:
separate PR even if so.

bdach reviewed (changes requested):
What do you mean when you say that tests don't work in the visual browser? The tests are headless, so they're not actually even visible in the browser, but I'm not sure if it's the cause or the bypass fix.

Skimmed through the code mostly as there are significant style issues to address first.

LittleEndu commented:
I marked the test as headless somewhere in the middle of this PR as I thought it was working but something changed and it no longer works in visual browser. So yeah, it being marked as headless is a fix not the cause.

LittleEndu committed:
Merge branch 'master' of https://github.com/ppy/osu into present-recommended

LittleEndu committed:
Unnest in SetUpSteps

LittleEndu committed:
Use GetRuleset

LittleEndu committed:
Fix typo

LittleEndu committed:
Other review suggestions

LittleEndu committed:
Fix comment

LittleEndu committed:
Fix early return check

LittleEndu committed:
Test presenting same beatmap more than once

peppy committed:
Merge branch 'master' into present-recommended

LittleEndu committed:
Merge branch 'master' of https://github.com/ppy/osu into present-recommended

 Conflicts:
	osu.Game/Screens/Select/SongSelect.cs

LittleEndu committed:
Merge remote-tracking branch 'origin/present-recommended' into present-recommended

bdach committed:
Improve test code quality & safety

bdach committed:
Do not query API for custom rulesets

bdach committed:
Simplify ruleset ordering

bdach commented:
So I spent an hour or two on this code to see if I can bring it to a mergeable state. I did fix up the test code, which was very haphazard in many places and I'm actually honestly surprised it ever passed CI, as in my runs through rider they would reliably fail every time due to prematurely accessing rulesets. In addition to that the registration of the API handling function needed to happen *before* instantiation of `OsuGame`.

Another issue I found is that the current code of `DifficultyRecommender` would attempt to fire off requests to fetch user profile data for custom rulesets, which I also fixed up. I also simplified one sort, but then I found an issue major enough as to warrant stopping further work on this.

The current code does not work well with the decoupled ruleset mechanism. Here's a simple reproduction scenario:

* Switch ruleset to taiko
* Open direct
* Present an osu! map
* The osu! map will be presented, but the difficulty selected will have star rating as if it were a taiko map

Because the decoupled ruleset gets transferred *after* the present happens, the recommender will try to select a difficulty too early.

As this is a major sticking point (to the point of threatening this PR's validity) I will not be continuing to work on this code further until this is resolved.

bdach reviewed (changes requested):
As above.

LittleEndu committed:
Add failing test

LittleEndu committed:
Fix recommender's incorrect usage of current ruleset

LittleEndu committed:
Add clarifying comment about pp choice

LittleEndu commented:
I was confused when you said "the recommender will try to select a difficulty too early" as recommender doesn't even deal with making a beatmap selection and shouldn't care about what ruleset is current. Turned out there was what I assume to be a copy paste oversight and the recommender was indeed using current ruleset. 

I assume you meant "selecting from its choices" and not "marking the beatmap as selected" as that's what song select and carousel do.

bdach commented on osu.Game/OsuGame.cs:
```diff
@@ -316,14 +316,12 @@ private void load()
         /// <param name="beatmap">The beatmap to select.</param>
         /// <param name="difficultyCriteria">
-        /// Optional predicate used to try and find a difficulty to select.
-        /// If omitted, this will try to present the first beatmap from the current ruleset.
+        /// Optional predicate used to filter which difficulties to select.
+        /// If omitted, this will try to present a recommended beatmap from the current ruleset.
         /// In case of failure the first difficulty of the set will be presented, ignoring the predicate.
         /// </param>
```
I'd reword this to something like
```csharp
        /// <param name="difficultyCriteria">Optional predicate used to narrow the set of difficulties to select from when presenting.</param>
        /// <remarks>
        /// Among items satisfying the predicate, the order of preference is:
        /// <list type="bullet">
        /// <item>beatmap with recommended difficulty, as provided by <see cref="DifficultyRecommender"/>,</item>
        /// <item>first beatmap from the current ruleset,</item>
        /// <item>first beatmap from any ruleset.</item>
        /// </list>
        /// </remarks>
```

bdach commented on osu.Game/OsuGame.cs:
```diff
@@ -341,16 +339,28 @@ public void PresentBeatmap(BeatmapSetInfo beatmap, Predicate<BeatmapInfo> diffic
+
+                // Try to select recommended beatmap
+                // This should give us a beatmap from current ruleset if there are any in our matched beatmaps
+                var selection = DifficultyRecommender.GetRecommendedBeatmap(beatmaps) ?? (
+                    // Fallback if a difficulty can't be recommended, maybe we are offline
+                    // First try to find a beatmap in current ruleset, otherwise use first beatmap
+                    beatmaps.FirstOrDefault(b => b.Ruleset.Equals(Ruleset.Value)) ?? beatmaps.First()
+                );
```
This reads super awkwardly. I'd split into three lines using `??=`.

LittleEndu replied:
This used to be split into multiple lines but that triggered CI errors. I'll see if using ``??=`` fixes that.

bdach commented on osu.Game/Screens/Select/DifficultyRecommender.cs:
```diff
@@ -72,6 +84,36 @@ private void calculateRecommendedDifficulties()
+
+        private IEnumerable<RulesetInfo> getBestRulesetOrder()
+        {
+            bestRulesetOrder ??= recommendedStarDifficulty.OrderByDescending(pair => pair.Value)
+                                                          .Select(pair => pair.Key)
+                                                          .ToList();
```
I'm not sure this aggressive caching is necessary, it's sorting a collection of at most 4 items. This isn't going to kill performance, and just ordering every time would get rid of this field.

peppy committed:
Merge branch 'master' into present-recommended

LittleEndu committed:
Merge branch 'master' of https://github.com/ppy/osu into present-recommended

 Conflicts:
	osu.Game/OsuGameBase.cs
	osu.Game/Screens/Select/DifficultyRecommender.cs
	osu.Game/Screens/Select/SongSelect.cs

LittleEndu committed:
Only calculate recommended SR once

LittleEndu committed:
Apply review suggestions

LittleEndu committed:
Remove no longer necessary force calculation

Is no longer necessary because recommender uses bindable state value changes, compared to when the test was written, and runs the state change once on load

peppy committed:
Merge branch 'master' into present-recommended

peppy commented on osu.Game/Screens/Select/DifficultyRecommender.cs:
```diff
@@ -47,21 +47,35 @@ private void load()
+                return;
+
+            // only query API for built-in rulesets
+            rulesets.AvailableRulesets.Where(ruleset => ruleset.ID <= 3).ForEach(rulesetInfo =>
```
Add this as a `const` inside of `ILegacyRuleset`

```csharp
        const int MAX_LEGACY_RULESET_ID = 3;
```

peppy commented on osu.Game/Screens/Select/DifficultyRecommender.cs:
```diff
@@ -75,12 +89,35 @@ private void calculateRecommendedDifficulties()
             });
         }
 
+        private IEnumerable<RulesetInfo> getBestRulesetOrder()
```
what does "best" mean here?

LittleEndu replied:
it means "best guess at what user wants to play". I'll add xmldoc to the function, but I can't come up with a better name.

peppy commented on osu.Game/Screens/Select/DifficultyRecommender.cs:
```diff
@@ -75,12 +89,35 @@ private void calculateRecommendedDifficulties()
             {
                 case APIState.Online:
-                    calculateRecommendedDifficulties();
+                    calculateRecommendedDifficulties(true);
```
probably want to clear any values on going offline

LittleEndu replied:
Not sure there are any to clear. After we have been online once we never need to change anything again.

peppy replied:
What if the user logs in as a different user?

LittleEndu replied:
User IDs are now checked, but still no action needs to be taken if the user logs out, unless we don't want to continue recommending the last logged in user's star rating. I think if somebody wants to play offline, they would still expect to get correct recommendations.

bdach replied:
Honestly it's debatable that it shouldn't be cleared. I can see an argument for clearing, too (say, having someone new play). We obviously won't know for sure which one it is, though, so I'm not sure I care much either way. This entire feature is almost too subtle to notice the existence of anyway.

LittleEndu replied:
Also wanted to note, lest I multi account, I don't think I can test this.

> too subtle to notice the existence of anyway

In song select it is very noticeable. in presentBeatmap's case, I can agree.

bdach replied:
I've never consciously been irked by (or even noted) the extra two or three arrow presses that it takes to select a difficulty in song select I want. Different strokes for different folks I suppose.

LittleEndu committed:
Merge branch 'master' of https://github.com/ppy/osu into present-recommended

LittleEndu committed:
Apply review suggestions

LittleEndu committed:
Document getBestRulesetOrder

LittleEndu committed:
Avoid calling ToList twice

LittleEndu committed:
Add test for recommending current ruleset

LittleEndu committed:
Merge remote-tracking branch 'origin/present-recommended' into present-recommended

peppy committed:
Merge branch 'master' into present-recommended

peppy commented on osu.Game/OsuGame.cs:
```diff
@@ -361,16 +363,29 @@ public void PresentBeatmap(BeatmapSetInfo beatmap, Predicate<BeatmapInfo> diffic
-                // Find first beatmap that matches our predicate.
-                var first = databasedSet.Beatmaps.Find(difficultyCriteria) ?? databasedSet.Beatmaps.First();
+                // Find beatmaps that match our predicate.
+                var beatmaps = databasedSet.Beatmaps.Where(b => difficultyCriteria?.Invoke(b) ?? true);
```
multiple enumeration

peppy committed:
Tidy up comments, code, and multiple linq enumeration

peppy committed:
Move and rename some fields for better readability

peppy committed:
Move from OsuGameBase to OsuGame

Also moves to a more suitable namespace.

peppy committed:
Allow recommender to potentially be null

peppy committed:
Remove unnecessary resolved recommender in test

peppy committed:
Refactor recommendation iteration code to read better

peppy commented:
@bdach would appreciate a quick sanity pass on this. i've made quite a few changes but am pretty confident.

bdach reviewed (approved):
seems reasonable

bdach committed:
Merge branch 'master' into present-recommended
